src/main/java/hudson/remoting/Channel.java

```diff
@@ -1314,10 +1316,12 @@ public synchronized void close(@CheckForNull Throwable diagnosis) throws IOExcep
      * (as opposed to executing {@link Callable}, which cannot have any reference to the context
      * of the remote {@link Channel}.
      */
-    public Object getProperty(Object key) {
+    @Override
+    public synchronized Object getProperty(Object key) {
```

Comment:
Agreed, it makes sense.
Or maybe even CopyOnWrite hashmap or ConcurrentHashMap to avoid explicit locks at all